watson/frames.py

```diff
@@ -68,88 +70,73 @@ def __contains__(self, frame):
-
         if isinstance(value, Frame):
-            frame = value
+            frame = self.new_frame(value.project, value.start, value.stop,
+                                   value.tags, value.updated_at, id=key)
         else:
-            frame = self.new_frame(*value)
+            frame = self.new_frame(*value[:5], id=key)
 
-        if isinstance(key, int):
-            self._rows[key] = frame
-        else:
-            frame = frame._replace(id=key)
-            try:
-                self._rows[self._get_index_by_id(key)] = frame
-            except KeyError:
-                self._rows.append(frame)
+        if key not in self:
+            self._keys.append(key)
 
-    def __delitem__(self, key):
+        super(Frames, self).__setitem__(key, frame)
         self.changed = True
 
-        if isinstance(key, int):
-            del self._rows[key]
-        else:
-            del self._rows[self._get_index_by_id(key)]
+    def __delitem__(self, key):
+        super(Frames, self).__delitem__(key)
+        self._keys.remove(key)
+        self.changed = True
 
-    def _get_index_by_id(self, id):
-        try:
-            return next(
-                i for i, v in enumerate(self['id']) if v.startswith(id)
-            )
-        except StopIteration:
-            raise KeyError("Frame with id {} not found.".format(id))
-
-    def _get_col(self, col):
-        index = HEADERS.index(col)
-        for row in self._rows:
-            yield row[index]
+    def move_to_end(self, key, last=True):
+        super(Frames, self).move_to_end(key, last)
+        self._keys.remove(key)
+        self._keys.insert(len(self._keys) if last else 0, key)
 
     def add(self, *args, **kwargs):
-        self.changed = True
         frame = self.new_frame(*args, **kwargs)
-        self._rows.append(frame)
+        self[frame.id] = frame
         return frame
 
-    def new_frame(self, project, start, stop, tags=None, id=None,
-                  updated_at=None):
-        if not id:
+    def new_frame(self, project, start, stop, tags=None, updated_at=None,
+                  id=None):
+        if id is None:
             id = uuid.uuid4().hex
-        return Frame(start, stop, project, id, tags=tags,
-                     updated_at=updated_at)
+
+        return Frame(id, project, start, stop, tags, updated_at)
 
     def dump(self):
-        return tuple(frame.dump() for frame in self._rows)
+        return tuple(frame.dump() for frame in self.values())
 
     def filter(self, projects=None, tags=None, span=None):
-        return (
-            frame for frame in self._rows
-            if (projects is None or frame.project in projects) and
-               (tags is None or any(tag in frame.tags for tag in tags)) and
-               (span is None or frame in span)
-        )
+        for frame in self.values():
+            if ((projects is None or frame.project in projects) and
+                    (tags is None or set(frame.tags).intersection(tags)) and
```

Comment:
The solution using `set` is actually slower:

```
In [1]: old_tags = list("abcdefghijkl")

In [2]: new_tags = list("ghijklmnopqr")

In [3]: %timeit any(tag in old_tags for tag in new_tags)
The slowest run took 10.96 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 731 ns per loop

In [4]: %timeit set(old_tags).intersection(new_tags)
The slowest run took 10.64 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.15 µs per loop
```

I only used it to be able to make the line shorter and fix a stupid flake 8 complaint about indentation. Though I think the slight performance loss is acceptable in this case.